test/run-drun/upgrade-remove-add-fields/version1.mo

```diff
@@ -0,0 +1,15 @@
+import Prim "mo:prim";
+
+actor {
+    stable var instance = {
+        var secondField = 0;
```

Comment:
Good test. This always scares me a little but I think it sound because the top-level mutable fields cannot be aliased so promoting to a supertype is ok. But do correct me if I'm wrong.

I think the upgade should not be allowed, for example, if the field was declared as
```
stable let instance = [ { ... } ].
```
with the contents in a first-class mutable location that could be aliased (and this written) from elsewhere.

Reply:
Maybe I misunderstand the question: I believe that fields cannot be directly aliased, only for example the record/object whose reference is in the variable. Whenever an object is promoted to super-type (removing fields from its original version), it should generally work even if the object is not directly referenced from an actor field, e.g. if part of an array type. If the object is aliased, an alias obtains a narrowed view of the same original object, and if an alias is updated, the object behind that alias is not altered but the variable replaced by a potentially new object. If  an object that is type-incompatible for the other alias would be assigned, it must be a different reference.

Example case (if I understood you correctly):

Version 0:
```
stable let instance [ { 
  var firstField  = 0;
  var secondField = 1;
} ];
stable let alias = instance;
```

Version 1:
```
stable let instance = [ { 
  var firstField  = 0;
  var secondField = 1;
} ];
stable let alias: [ { // Dropping firstField only on alias
   var secondField = 1;  
} = ...;
```
I think this upgrade should be allowed.
Even if I change the following, this would be okay:

```
alias[0] := { var secondField = 2 };
```
Now the `alias` variable is separated from `instance`.

But, probably, I misunderstood your question. Maybe I miss some relevant point?

Reply:
And yes fields cannot be aliases so you can drop them in strictly positive/covariant positions.

Reply:
Oh yes, now I see. Mutable arrays must be invariant. I will double check.

Reply:
Indeed, I had to fix memory compatibility check. It was too permissive with objects, variants, and service types that are part of mutable variable/array. See https://github.com/dfinity/motoko/commit/7d71dafeea24bd049ead0d69dabadefc69a0aa3d